pinot-common/src/main/java/org/apache/pinot/sql/parsers/rewriter/CompileTimeFunctionsInvoker.java

```diff
@@ -61,38 +64,40 @@ protected static Expression invokeCompileTimeFunctionExpression(@Nullable Expres
     List<Expression> operands = function.getOperands();
     int numOperands = operands.size();
     boolean compilable = true;
+    ColumnDataType[] argumentTypes = new ColumnDataType[numOperands];
+    Object[] arguments = new Object[numOperands];
     for (int i = 0; i < numOperands; i++) {
       Expression operand = invokeCompileTimeFunctionExpression(operands.get(i));
-      if (operand.getLiteral() == null) {
+      operands.set(i, operand);
+      Literal literal = operand.getLiteral();
+      if (compilable && literal != null) {
+        RequestUtils.getLiteralTypeAndValue(literal, argumentTypes, arguments, i);
+      } else {
         compilable = false;
       }
-      operands.set(i, operand);
     }
-    if (compilable) {
-      String canonicalName = FunctionRegistry.canonicalize(function.getOperator());
-      FunctionInfo functionInfo = FunctionRegistry.lookupFunctionInfo(canonicalName, numOperands);
-      if (functionInfo != null) {
-        Object[] arguments = new Object[numOperands];
-        for (int i = 0; i < numOperands; i++) {
-          arguments[i] = RequestUtils.getLiteralValue(function.getOperands().get(i).getLiteral());
-        }
-        try {
-          FunctionInvoker invoker = new FunctionInvoker(functionInfo);
-          Object result;
-          if (invoker.getMethod().isVarArgs()) {
-            result = invoker.invoke(new Object[] {arguments});
-          } else {
-            invoker.convertTypes(arguments);
-            result = invoker.invoke(arguments);
-          }
-          return RequestUtils.getLiteralExpression(result);
-        } catch (Exception e) {
-          throw new SqlCompilationException(
-              "Caught exception while invoking method: " + functionInfo.getMethod() + " with arguments: "
-                  + Arrays.toString(arguments), e);
-        }
+    if (!compilable) {
+      return expression;
+    }
+    String canonicalName = FunctionRegistry.canonicalize(function.getOperator());
+    FunctionInfo functionInfo = FunctionRegistry.lookupFunctionInfo(canonicalName, argumentTypes);
```

Comment:
I'm wondering if we might run into some type conversion issues here. For instance, a polymorphic scalar comparison function may implement support for comparing two arguments of the same type (see https://github.com/apache/pinot/pull/13711 for instance). Earlier with the argument count based function lookup, if the two arguments were of a different type, the scalar comparison function invocation might still have worked based on the type conversion in [FunctionInvoker](https://github.com/apache/pinot/blob/835949b64529b8e5c0bb49474cdec2e17c9050f7/pinot-common/src/main/java/org/apache/pinot/common/function/FunctionInvoker.java#L94). But now, each such polymorphic scalar function implementation would need to either account for all these different type combinations or else fall back to a default implementation and then rely on the `FunctionInvoker`'s type conversion in case of heterogeneous argument types.

Reply:
The scalar function should take care of all possible combinations of argument types, similar to the backward compatible logic in `PolymorphicComparisonScalarFunction` when argument type is different